src/Haskell/Ide/Engine/Plugin/Ormolu.hs

```diff
@@ -34,24 +44,69 @@ ormoluDescriptor plId = PluginDescriptor
 
 
 provider :: FormattingProvider
-provider _contents _uri _typ _opts =
 #if __GLASGOW_HASKELL__ >= 806
-  case _typ of
-    FormatRange _ -> return $ IdeResultFail (IdeError PluginError (T.pack "Selection formatting for Ormolu is not currently supported.") Null)
-    FormatText -> pluginGetFile _contents _uri $ \file -> do
-        opts <- lookupComponentOptions file
-        let opts' = map DynOption $ filter exop $ join $ maybeToList $ componentOptions <$> opts
-            conf  = Config opts' False False True False
-        result <- liftIO $ try @OrmoluException (ormolu conf file (T.unpack _contents))
-
-        case result of
-          Left  err -> return $ IdeResultFail (IdeError PluginError (T.pack $ "ormoluCmd: " ++ show err) Null)
-          Right new -> return $ IdeResultOk [TextEdit (fullRange _contents) new]
-  where
-    exop s =
-      "-X" `isPrefixOf` s
-      || "-fplugin=" `isPrefixOf` s
-      || "-pgmF=" `isPrefixOf` s
+{-# LANGUAGE BlockArguments #-}
+provider contents uri typ _ = pluginGetFile contents uri $ \fp -> do
+  opts <- lookupComponentOptions fp
+  let cradleOpts =
+        map DynOption
+          $   filter exop
+          $   join
+          $   maybeToList
+          $   componentOptions
+          <$> opts
+
+      fromDyn tcm _ () =
+        let
+          df = getDynFlags tcm
+          pp =
+            let p = D.sPgm_F $ D.settings df
+            in  if null p then [] else ["-pgmF=" <> p]
+          pm = map (("-fplugin=" <>) . moduleNameString) $ D.pluginModNames df
+          ex = map (("-X" <>) . show) $ S.toList $ D.extensionFlags df
+        in
+          return $ map DynOption $ pp <> pm <> ex
+  fileOpts <- ifCachedModuleAndData fp cradleOpts fromDyn
+  let
+    conf o = Config o False False True False
+    fmt :: T.Text -> [DynOption] -> IdeM (Either OrmoluException T.Text)
+    fmt cont o =
+      liftIO $ try @OrmoluException (ormolu (conf o) fp $ T.unpack cont)
+
+  case typ of
+    FormatText -> ret (fullRange contents) <$> fmt contents cradleOpts
+    FormatRange r ->
+      let
+        txt = T.lines $ extractRange r contents
+        lineRange (Range (Position sl _) (Position el _)) =
+          Range (Position sl 0) $ Position el $ T.length $ last txt
+        fixLine t = if T.all isSpace $ last txt then t else T.init t
+        unStrip ws new =
+          fixLine $ T.unlines $ map (ws `T.append`) $ T.lines new
+        mStrip = case txt of
+          (l : _) ->
+            let ws = fst $ T.span isSpace l
+            in  (,) ws . T.unlines <$> traverse (T.stripPrefix ws) txt
+          _ -> Nothing
+        err = return $ IdeResultFail
+          (IdeError
+            PluginError
+            (T.pack
+              "You must format a whole block of code. Ormolu does not support arbitrary ranges."
```

Comment:
What user experience does this result in? A pop-up window?  This is likely to be a fairly common experience, so should be non-threatening, just a notice.

Another possible approach is to find the enclosing block that *can* be formatted, and process that.  What does the brittany plugin do in this case?  I wonder if that logic should be captured in a library, for use in all RangeFormatting requests, so we have a standard behaviour across hie.

Reply:
I use nvim and coc so I don't know how it looks to most users. Brittany will often produce an error, in similar situations. The point of this line is just to provide a more helpful error than what we would usually get `unable to parse`.

Reply:
I am happy with the message, and its contents. I just want to make sure the end user experience makes sense. But it can be tweaked easily enough, so its not a blocker.